File: backend/app/models/debrief.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class ConversationStats(BaseModel):
    talk_listen_ratio: float
    question_count: int
    open_question_count: int = 0
    closed_question_count: int = 0
    interruption_count: int
    average_turn_offset_ms: int = 0
    turn_offset_series: list[dict[str, Any]] = Field(default_factory=list)
    session_duration_minutes: float
    user_speech_duration_minutes: float
    other_speech_duration_minutes: float = 0.0
    estimated_wpm: float
    energy_score: int = 0
    energy_axes: list[float] = Field(default_factory=lambda: [0.0, 0.0, 0.0])
    energy_series_user: list[float] = Field(default_factory=list)
    energy_series_other: list[float] = Field(default_factory=list)
    lsm_score: float = 0.0
    lsm_dimensions_user: dict[str, float] = Field(default_factory=dict)
    lsm_dimensions_reference: dict[str, float] = Field(default_factory=dict)
    total_word_count: int = 0
    unique_word_count: int = 0
    vocabulary_richness: float = 0.0
    filler_counts: list[dict[str, Any]] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Debrief(BaseModel):
    id: UUID
    session_id: str
    user_id: str
    created_at: datetime
    observation: str
    pattern_to_reduce: str
    thing_to_try_next: str
    stats: ConversationStats
    transcript: str | None = None


class SessionResponse(BaseModel):
    debrief: Debrief
    used_this_month: int
    remaining: int

File: app/src/models/debrief.ts
export interface ConversationStats {
  talkListenRatio: number;
  questionCount: number;
  openQuestionCount: number;
  closedQuestionCount: number;
  interruptionCount: number;
  averageTurnOffsetMs: number;
  turnOffsetSeries: { t: string; ms: number }[];
  sessionDurationMinutes: number;
  userSpeechDurationMinutes: number;
  otherSpeechDurationMinutes: number;
  estimatedWpm: number;
  energyScore: number;
  energyAxes: number[];
  energySeriesUser: number[];
  energySeriesOther: number[];
  lsmScore: number;
  lsmDimensionsUser: Record<string, number>;
  lsmDimensionsReference: Record<string, number>;
  totalWordCount: number;
  uniqueWordCount: number;
  vocabularyRichness: number;
  fillerCounts: FillerCount[];
  metadata: Record<string, unknown>;
}

export interface DebriefCard {
  id: string;
  sessionId: string;
  userId: string;
  createdAt: string;
  observation: string;
  patternToReduce: string;
  thingToTryNext: string;
  stats: ConversationStats;
  transcript?: string | null;
}

export interface UsageSummary {
  usedThisMonth: number;
  remaining: number;
  resetsAt: string;
}

export interface SessionResponse {
  debrief: DebriefCard;
  usedThisMonth: number;
  remaining: number;
}

export interface ConversationSummary {
  id: string;
  title: string;
  createdAt: string;
  durationMinutes: number;
  talkListenPercent: number;
  questionCount: number;
  interruptionCount: number;
  energyScore: number;
  lsmScore: number;
  vocabularyRichness: number;
  tone: string;
  note: string;
}

export interface FillerCount {
  phrase: string;
  count: number;
}

export interface ProgressWeekSummary {
  weekKey: string;
  label: string;
  startsAt: string;
  endsAt: string;
  conversationCount: number;
  totalMinutes: number;
  dailyMinutes: number[];
  dailyQuestions: number[];
  dailyOpenQuestions: number[];
  dailyClosedQuestions: number[];
  dailyInterruptions: number[];
  averageSessionMinutes: number;
  talkListenPercent: number;
  averageQuestions: number;
  totalQuestions: number;
  totalOpenQuestions: number;
  totalClosedQuestions: number;
  interruptionCount: number;
  averageTurnOffsetMs: number;
  dailyTurnOffsets: (number | null)[];
  averageWpm: number;
  energyScore: number;
  energyAxes: number[];
  lsmAverage: number;
  vocabularyUniqueWords: number;
  vocabularyTotalWords: number;
  vocabularyRichness: number;
  topFillers: FillerCount[];
  wins: string[];
  nudges: string[];
  conversations: ConversationSummary[];
}

export interface ProgressSummary {
  generatedAt: string;
  currentWeekIndex: number;
  weeks: ProgressWeekSummary[];
}

export interface ProfileSummary {
  totalConversations: number;
  totalMinutes: number;
  averageQuestions: number;
  talkListenPercent: number;
  usedThisMonth: number;
  remaining: number;
  resetsAt: string;
}

export interface BillingStatus {
  plan: 'free' | 'pro';
  status: string;
  isPro: boolean;
  freeConversationsRemaining: number;
  currentPeriodEnd?: string | null;
  trialEnd?: string | null;
  cancelAtPeriodEnd: boolean;
  stripeConfigured: boolean;
  checkoutAvailable: boolean;
  portalAvailable: boolean;
}

export interface AccountExport {
  exportedAt: string;
  userId: string;
  profile: ProfileSummary;
  settings: UserSettings;
  billing: BillingStatus;
  debriefs: DebriefCard[];
}

export interface ReflectMessage {
  role: 'assistant' | 'user';
  content: string;
}

export type WeeklySummaryDay = 'sunday' | 'monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday';
export type WeeklySummaryTime = 'early_morning' | 'morning' | 'midday' | 'afternoon' | 'evening' | 'night';
export type CoachingTone = 'warm_reflective' | 'direct_practical' | 'curious_gentle';
export type CoachingDepth = 'quick' | 'balanced' | 'deep';

export interface UserSettings {
  notificationsEnabled: boolean;
  weeklySummaryDay: WeeklySummaryDay;
  weeklySummaryTime: WeeklySummaryTime;
  reflectionReminders: boolean;
  productUpdates: boolean;
  saveTranscripts: boolean;
  includeTranscriptInReflect: boolean;
  coachingTone: CoachingTone;
  coachingDepth: CoachingDepth;
}
